ngraph_bridge/ngraph_capture_variables.cc

```diff
@@ -201,10 +215,19 @@ Status CaptureVariables(Graph* graph, const std::set<string> skip_these_nodes) {
         }
 
         replaced_nodes.push_back(node);
-      } else if (node->type_string() == "PrefetchDataset") {
-        // Collect the prefetch_node so that we can add
-        // the NGraphWriteToDevice Op after this one
-        prefetch_node = node;
+      } else if (node->type_string() == "MakeIterator") {
+        if (std::getenv(NGraphPrefetchSharedResouce::NGRAPH_TF_USE_PREFETCH) !=
+            nullptr) {
+          // Recursively go over the inputs of MakeIterator and
+          // collect the first PrefetchDataset op found in the
+          // path so that we can add
+          // the NGraphWriteToDevice Op after this one
+          std::vector<Node*> in_nodes;
+          for (auto edge : node->in_edges()) {
+            in_nodes.push_back(edge->src());
+          }
+          prefetch_node = FindPrefetch(in_nodes);
```

Comment:
If there exists a `PrefetchDataset` feeding into a `MakeIterator` return that, or else, look at the inputs of the inputs etc

But AFAIR, we were only looking for `PrefetchDataset-->MakeIterator` kind of linkages. In fact exactly 1 `PrefetchDataset` feeding into a `MakeIterator`. If we do not find that we throw an error.

I could be remembering the discussion incorrectly.

Reply:
Yes, we did discuss that but we also want to support both the examples at hand (axpy and resnet). For resnet what we had discussed would work but not for axpy, hence the change in the algorithm.

Reply:
Ok, for axpy we have: `PrefetchDataset-->OptimizeDataset-->etc--> MakeIterator `

We need to throw an error if `in_nodes.size() == 0`. Because imagine a chain with no prefetchdataset, then it would be in an infinite recursion i think

Reply:
The algorithm has changed per discussion, there is no recursion anymore.